keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,549 @@
+- Starting Jobs based on Ready will be at the ReplicatedJob level not at the Job level.
+  - ReplicatedJobs will be classified as ready if all jobs are ready.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I have one ReplicatedJob that functions as the driver and a series of worker ReplicatedJobs.  
+I want the driver to be ready before I start the worker replicated jobs.  
+This is useful for HPC/AI/ML workloads as some frameworks have a driver pod that must be ready before the workers are started.  
+
+The API below would fit my purpose.
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: driver-ready-worker-start
+spec:
+  startupPolicy:
+  - name: driver
+    status: Ready
+  replicatedJobs:
+  - name: driver
+    replicas: 1
+    template:
+      spec:
+        # Set backoff limit to 0 so job will immediately fail if any pod fails.
+        backoffLimit: 0 
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: driver
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10000
+  - name: workers
+    replicas: 1
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+```
+
+#### Story 2
+
+As a user, I have one ReplicatedJob that I want to finish before going on to the next worker.
+
+This can be useful as a simple way to target jobs to run in order.
+
+An API that fits my purpose is below.
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: driver-ready-worker-start
+spec:
+  startupPolicy:
+  - name: driver
+    status: Succeeded
+  replicatedJobs:
+  - name: first
+    replicas: 1
+    template:
+      spec:
+        # Set backoff limit to 0 so job will immediately fail if any pod fails.
+        backoffLimit: 0 
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: driver
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 100
+  - name: worker
+    replicas: 2
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+```
+
+#### Story 3
+
+As a user I only need a percentage of my targeted replicated job to succeed before going on to the next job.
```

Comment:
Interesting idea, what is the use case for this though? What applications/frameworks require a percentage threshold like this?

Reply:
+1, what is the use case?

Reply:
I'm not sure anymore.  

I was thinking of https://github.com/kubernetes/kubernetes/issues/117758 and [Success/Failure Policy](https://github.com/kubernetes/kubernetes/issues/117600) in the job controller.  

In this case, you could have certain jobs be considered successful or failed based on percentage of indexes that passed/failed.  I thought maybe it could be a good idea to expand that to JobSet where you have certain percentage of replicatedJobs match a targerted status.  

I was thinking that if you had a bunch of replicas of the same ReplicatedJob, then wouldn't you also want to make sure that they all succeed.  But then it could be possible for some of them to fail and others to pass.  

I had a little too much on my last days in my last company so I was feeling inspired.  

Maybe this is YAGNI and we move this to a non goal?

Reply:
+1 YAGNI.